Objects/unicodeobject.c

```diff
@@ -2438,25 +2458,44 @@ PyUnicode_Export(PyObject *unicode, int32_t requested_formats,
         if (bytes == NULL) {
             return -1;
         }
-        ucs4 = (Py_UCS4*)PyBytes_AS_STRING(bytes);
 
-        int32_t res = unicode_export(bytes, view,
-                                     len, ucs4,
-                                     4, BUFFER_UCS4, PyUnicode_FORMAT_UCS4);
-        Py_DECREF(bytes);
-        return res;
+        return unicode_export_bytes(bytes, view,
+                                    4, BUFFER_UCS4, PyUnicode_FORMAT_UCS4);
     }
 
     // Encode UCS1, UCS2 or UCS4 to UTF-8
     if (requested_formats & PyUnicode_FORMAT_UTF8) {
         Py_ssize_t nbytes;
         const char *utf8 = PyUnicode_AsUTF8AndSize(unicode, &nbytes);
-        if (utf8 == NULL) {
-            return -1;
+        if (utf8 != NULL) {
+            return unicode_export(unicode, view,
+                                  nbytes, utf8,
+                                  1, "B", PyUnicode_FORMAT_UTF8);
         }
-        return unicode_export(unicode, view,
-                              nbytes, utf8,
-                              1, "B", PyUnicode_FORMAT_UTF8);
+        if (PyErr_ExceptionMatches(PyExc_UnicodeEncodeError)) {
+            PyErr_Clear();
+            PyObject *bytes = _PyUnicode_AsUTF8String(unicode, "surrogatepass");
+            if (bytes == NULL) {
+                return -1;
+            }
+            len = PyBytes_GET_SIZE(bytes);
+
+            // Copy to add a NULL character
+            PyObject *bytes2 = PyBytes_FromStringAndSize(NULL, len + 1);
```

Comment:
> _PyUnicode_AsUTF8String() returns a null terminated bytes object, so this is not needed.

If I remove `bytes2`, test_capi.test_unicode fails with an assertion error since the string doesn't end with a NULL character. I don't see where _PyUnicode_AsUTF8String() writes a NULL character.

Reply:
Every bytes object should be NUL-terminated. If it is not, this is a bug.

Maybe `_PyUnicode_AsUTF8String()` truncates an overallocated bytes object but forget to write NUL?

Reply:
Extract of `unicode_encode_utf8()`:

```c
    if (PyUnicode_UTF8(unicode))
        return PyBytes_FromStringAndSize(PyUnicode_UTF8(unicode),
                                         PyUnicode_UTF8_LENGTH(unicode));
```

It doesn't overallocate in any code path. It just omits the trailing NULL character.

Reply:
Should not `PyBytes_FromStringAndSize()` (and any other PyBytes constructor) write a null byte past the end of the array?

BTW, this is not "NULL character", this is called "null character". It also has name NUL in the ASCII table. NULL is confusing in this context, as it means a null pointer.

Reply:
Oh, so maybe it's the `PyBytesWriter` API which doesn't write the NUL byte at the end?

Reply:
Ok, I found the bug and I removed `bytes2` copy. It's no longer needed.